src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +30,449 @@
 public class EntityAIWorkDeliveryman extends AbstractEntityAIInteract<JobDeliveryman>
 {
     /**
+     * Min distance the worker should have to the warehouse to make any decisions.
+     */
+    private static final int MIN_DISTANCE_TO_WAREHOUSE = 5;
+
+    /**
+     * Walking speed double at this level.
+     */
+    private static final double WALKING_SPEED_MULTIPLIER = 25;
+
+    /**
+     * Min distance to chest to take something out of it.
+     */
+    private static final int MIN_DISTANCE_TO_CHEST = 2;
+
+    /**
+     * The base movement speed of the deliveryman.
+     */
+    private static final double BASE_MOVEMENT_SPEED = 0.2D;
+
+    /**
+     * Delay in ticks between every inventory operation.
+     */
+    private static final int DUMP_AND_GATHER_DELAY  = 10;
+
+    /**
+     * Warehouse the deliveryman is assigned to.
+     */
+    private BuildingWareHouse wareHouse = null;
+
+    /**
+     * List of itemStacks the deliveryman will deliver to the requesting building.
+     */
+    private List<ItemStack> itemsToDeliver = new ArrayList<>();
+
+    /**
+     * Next target the deliveryman should gather stuff at.
+     */
+    private BlockPos gatherTarget = null;
+
+    /**
+     * Amount of stacks left to gather from the inventory at the gathering step.
+     */
+    private int currentSlot = 0;
+
+    /**
+     * Amount of stacks the worker already kept in the current gathering process.
+     */
+    private List<ItemStorage> alreadyKept = new ArrayList<>();
+
+
+    /**
      * Initialize the deliveryman and add all his tasks.
      *
      * @param deliveryman the job he has.
      */
     public EntityAIWorkDeliveryman(@NotNull final JobDeliveryman deliveryman)
     {
         super(deliveryman);
+        super.registerTargets(
+                /**
+                 * Check if tasks should be executed.
+                 */
+                new AITarget(this::checkIfExecute, IDLE),
+                new AITarget(IDLE, () -> START_WORKING),
+                new AITarget(START_WORKING, this::checkWareHouse),
+                new AITarget(PREPARE_DELIVERY, this::prepareDelivery),
+                new AITarget(GATHER_IN_WAREHOUSE, this::gatherItemsFromWareHouse),
+                new AITarget(DELIVERY, this::deliver),
+                new AITarget(GATHERING, this::gather),
+                new AITarget(DUMPING, this::dump)
+
+        );
+        worker.setSkillModifier(2 * worker.getCitizenData().getEndurance() + worker.getCitizenData().getCharisma());
+        worker.setCanPickUpLoot(true);
+    }
+
+    /**
+     * Gather items from a random hut which the hut doesn't need.
+     * @return the next state to go to.
+     */
+    public AIState gather()
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this method 6, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "gather" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)